src/main/java/io/cryostat/reports/MemoryCachingReportsService.java

```diff
@@ -108,4 +110,20 @@ public Uni<Map<String, AnalysisResult>> reportFor(ActiveRecording recording) {
     public Uni<Map<String, AnalysisResult>> reportFor(String jvmId, String filename) {
         return reportFor(jvmId, filename, r -> true);
     }
+
+    @Override
+    public boolean keyExists(ActiveRecording recording) {
+        String key = ReportsService.key(recording);
+        return (quarkusCache || memoryCache)
+                && !Objects.isNull(activeCache.as(CaffeineCache.class).getIfPresent(key))
+                && !delegate.keyExists(recording);
+    }
+
+    @Override
+    public boolean keyExists(String jvmId, String filename) {
+        String key = recordingHelper.archivedRecordingKey(jvmId, filename);
+        return (quarkusCache || memoryCache)
+                && !Objects.isNull(archivedCache.as(CaffeineCache.class).getIfPresent(key))
+                && !delegate.keyExists(jvmId, filename);
```

Comment:
Could you explain the logic here? I would think that this class considers that the key exists either if (it itself is enabled AND it itself has it in its cache), OR if the delegate has it. So that would be like `((quarkusCache && memoryCache) && archivedCache.as(CaffeineCache.class).keySet().contains(key)) || delegate.keyExists(jvmId, filename)`

Instead we have an inverted condition on the delegate, and also the "is the cache enabled" condition is different from in the `reportFor` methods.

Reply:
I made a mistake here when we added the check for the delegate, I've fixed it now. Good catch

Reply:
This still looks different than I expected - `quarkusCache || memoryCache` vs `quarkusCache && memoryCache`. I think both need to be true, since `quarkusCache` indicates whether the overall caching layer is enabled and `memoryCache` indicates whether this particular cache should be enabled. The other references to these values within the class use `!quarkusCache || !memoryCache` as an early-return condition to bail out, so the inverse of that condition is what we want here. `! (!p || !q) == (p && q)`